pkg/virtctl/vm/vm.go

```diff
@@ -433,229 +108,35 @@ func getDryRunOption(dryRun bool) []string {
-			return fmt.Errorf("error invalid arguments - VirtualMachine name or file must be provided")
-		}
-
-		if outputFormat != YAML && outputFormat != JSON {
-			return fmt.Errorf("error not supported output format defined: %s", outputFormat)
-		}
-
-		return nil
-	}
-}
-
-func readVMFromFile(filePath string) (*v1.VirtualMachine, error) {
-	vm := &v1.VirtualMachine{}
-
-	readFile, err := os.ReadFile(filePath)
-	if err != nil {
-		return nil, fmt.Errorf("error reading file %+w", err)
-	}
-
-	err = yaml.Unmarshal(readFile, vm)
-	if err != nil {
-		return nil, fmt.Errorf("error decoding VirtualMachine %+w", err)
-	}
-
-	return vm, nil
-}
-
-func applyOutputFormat(outputFormat string, expandedVm *v1.VirtualMachine) (string, error) {
-	var formatedOutput []byte
-	var err error
-
-	switch outputFormat {
-	case JSON:
-		formatedOutput, err = json.MarshalIndent(expandedVm, "", " ")
-	case YAML:
-		formatedOutput, err = yaml.Marshal(expandedVm)
-	}
-
-	if err != nil {
-		return "", err
-	}
-
-	return string(formatedOutput), nil
+type vmOptions struct {
+	vmiName      string
+	namespace    string
+	dryRunOption []string
+	virtClient   kubecli.KubevirtClient
 }
 
-func expandVirtualMachine(namespace string, virtClient kubecli.KubevirtClient, o *Command) error {
-	var expandedVm *v1.VirtualMachine
+func BeforeRun(args []string, clientConfig clientcmd.ClientConfig) (*vmOptions, error) {
+	vars := vmOptions{}
 	var err error
 
-	if vmName != "" {
-		expandedVm, err = virtClient.VirtualMachine(namespace).GetWithExpandedSpec(vmName)
-		if err != nil {
-			return fmt.Errorf("error expanding VirtualMachine - %s in namespace - %s: %w", vmName, namespace, err)
-		}
-	} else {
-		vm, err := readVMFromFile(filePath)
-		if err != nil {
-			return err
-		}
-
-		expandedVm, err = virtClient.ExpandSpec(namespace).ForVirtualMachine(vm)
-		if err != nil {
-			return fmt.Errorf("error expanding VirtualMachine - %s in namespace - %s: %w", vm.Name, namespace, err)
-		}
-	}
-
-	output, err := applyOutputFormat(outputFormat, expandedVm)
-	if err != nil {
-		return err
-	}
-
-	o.cmd.Print(output)
-	return nil
-}
-
-func (o *Command) Run(args []string) error {
 	if len(args) != 0 {
-		vmiName = args[0]
+		vars.vmiName = args[0]
 	}
 
-	namespace, _, err := o.clientConfig.Namespace()
+	vars.namespace, _, err = clientConfig.Namespace()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(o.clientConfig)
+	vars.virtClient, err = kubecli.GetKubevirtClientFromClientConfig(clientConfig)
 	if err != nil {
-		return fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
+		return nil, fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
 	}
 
-	dryRunOption := getDryRunOption(dryRun)
-	if len(dryRunOption) > 0 && dryRunOption[0] == metav1.DryRunAll {
+	vars.dryRunOption = getDryRunOption(dryRun)
+	if len(vars.dryRunOption) > 0 && vars.dryRunOption[0] == metav1.DryRunAll {
 		fmt.Printf("Dry Run execution\n")
```

Comment:
Is this really applicable to all vm cmds? What happens when you pass the dry run option when the cmd does not support it? I'd vote to separate this out and only using it where applicable (modifications being made).